core/dbt/deps/tarball.py

```diff
@@ -0,0 +1,249 @@
+import os
+import functools
+import hashlib
+import tarfile
+from tempfile import NamedTemporaryFile
+from typing import Optional, Union
+from dbt.clients import system
+from dbt.deps.base import PinnedPackage, UnpinnedPackage, get_downloads_path
+from dbt.exceptions import (
+    DependencyException, 
+    package_structure_malformed, 
+    package_sha1_fail
+)
+from dbt.events.functions import fire_event
+from dbt.events.types import Sha1ChecksumPasses, UntarProjectRoot
+from dbt.config import Project
+from dbt.contracts.project import (
+    ProjectPackageMetadata,
+    TarballPackage,
+)
+from dbt.utils import _connection_exception_retry as connection_exception_retry
+
+
+TARFILE_MAX_SIZE = 1 * 1e+6  # limit tarfiles to 1mb
+
+
+class TarballPackageMixin:
+    def __init__(self, tarball: str) -> None:
+        super().__init__()
+        self.tarball = tarball
+
+    @property
+    def name(self):
+        return self.tarball
+
+    def source_type(self) -> str:
+        return 'tarball'
+
+
+class TarballPinnedPackage(TarballPackageMixin, PinnedPackage):
+    def __init__(
+        self,
+        tarball: Union[str, tarfile.TarFile],
+        sha1: Optional[str] = None,
+        subdirectory: Optional[str] = None,
+    ) -> None:
+        super().__init__(tarball)
+        self.sha1 = sha1
+        self.subdirectory = subdirectory
+        # self.tarfile_bin = self.get_tarfile()
+        # self.get_tarfile()
+        # self.tar_dir_name = self.resolve_tar_dir()
+
+    def get_version(self):
+        return None
+
+    def nice_version_name(self):
+        return '<tarball @ {}>'.format(self.tarball)
+
+    def get_tar_size(self):
+        return sum(x.size for x in self.tarfile_bin.getmembers())
+
+    def file_sha1(self, fp_in: str):
+        if type(fp_in) is str:
+            with open(fp_in, 'rb') as fp:
+                return hashlib.sha1(fp.read()).hexdigest()
+        elif hasattr(fp_in, 'read'):
+            return hashlib.sha1(fp_in.read()).hexdigest()
+        else:
+            raise ValueError()
+
+    def get_tarfile(self):
+        def validate_checksum(self, filepath):
+            if self.sha1:
+                checksum = self.file_sha1(filepath)
+                if not checksum == self.sha1:
+                    package_sha1_fail(checksum, self.sha1)
+
+        def validate_tarfile(self, filepath):
+            if not tarfile.is_tarfile(filepath):
+                msg = f"{filepath} is not a valid tarfile."
+                raise DependencyException(msg)
+
+        def handle_local_tar(self):
+            # pass through local tarfile path primarily for testing
+            validate_checksum(self, self.tarball)
+            validate_tarfile(self, self.tarball)
+            tar = tarfile.open(self.tarball, "r:gz")
+            self.tarfile_bin = tar
```

Comment:
This is duplicated by part of the code in `handle_remote_tar` (#L102-L105).  Combining this will also remove the need for the nested functions as they won't be called more than once.

Reply:
Ok - updated accordingly, see 1dcd97bbc8176060d10c849ed584550ab3f410bd